File: src/planets/transportColonists.ts
import type { Ship } from "../data/types/ship.ts";
import type { Planet } from "../data/types/planet.ts";
import type { Wallet } from "../data/types/wallet.ts";
import type { PlanetOwnershipEntry } from "../data/types/planetOwnershipEntry.ts";
import type { TransportColonistsResult } from "../data/types/transportColonistsResult.ts";
import { COLONIST_TRANSPORT_COST } from "../data/constants/planetOwnership.ts";

// Colonist-Driven Production (planet-ownership.md). Requires the ship be
// docked at destinationPlanet -- without this, the ship parameter would
// have no enforced purpose, and "colonists are transported" would be true
// in name only. Colonists have no separate origin/supply -- abstracted as
// "arranged via credits" once docked, not carried from a source planet or
// limited pool (a deliberate simplification, per the design entry).
export function transportColonists(
  ship: Ship,
  destinationPlanet: Planet,
  quantity: number,
  wallet: Wallet,
  currentOwnershipEntry: PlanetOwnershipEntry,
): TransportColonistsResult {
  if (ship.currentPlanetId !== destinationPlanet.id) {
    return { success: false, reason: "ship must be docked at the destination planet to transport colonists" };
  }
  if (quantity <= 0) {
    return { success: false, reason: "quantity must be positive" };
  }

  const cost = quantity * COLONIST_TRANSPORT_COST;
  if (wallet.credits < cost) {
    return { success: false, reason: `insufficient funds: need ${cost}, have ${wallet.credits}` };
  }

  return {
    success: true,
    updatedWallet: { ...wallet, credits: wallet.credits - cost },
    updatedOwnershipEntry: {
      ...currentOwnershipEntry,
      colonistCount: currentOwnershipEntry.colonistCount + quantity,
    },
  };
}

File: unity/ProfitableCore/Simulation/TransportColonists.cs
using Profitable.Core.Constants;
using Profitable.Core.Schema;

namespace Profitable.Core.Simulation;

// Ports src/planets/transportColonists.ts. Requires the ship be docked at
// destinationPlanet -- without this, the ship parameter would have no
// enforced purpose. Colonists have no separate origin/supply -- abstracted
// as "arranged via credits" once docked, not carried from a source planet
// or limited pool (a deliberate simplification).
//
// Retroactive removal (2026-08-04): this used to also copy through
// CitadelLevel/OwnedByPlayerId on the updated entry -- both fields removed
// along with Citadels, see planet-ownership.md's own retroactive note.
// ColonistCount handling is unaffected.
public static class ColonistTransporter
{
    public static TransportColonistsResult TransportColonists(
        Ship ship,
        Planet destinationPlanet,
        int quantity,
        Wallet wallet,
        PlanetOwnershipEntry currentOwnershipEntry)
    {
        if (ship.CurrentPlanetId != destinationPlanet.Id)
        {
            return new TransportColonistsRejected { Reason = "ship must be docked at the destination planet to transport colonists" };
        }
        if (quantity <= 0)
        {
            return new TransportColonistsRejected { Reason = "quantity must be positive" };
        }

        var cost = quantity * PlanetOwnershipConstants.ColonistTransportCost;
        if (wallet.Credits < cost)
        {
            return new TransportColonistsRejected { Reason = $"insufficient funds: need {cost}, have {wallet.Credits}" };
        }

        return new TransportColonistsSucceeded
        {
            UpdatedWallet = new Wallet { PlayerId = wallet.PlayerId, Credits = wallet.Credits - cost },
            UpdatedOwnershipEntry = new PlanetOwnershipEntry
            {
                ColonistCount = currentOwnershipEntry.ColonistCount + quantity,
            },
        };
    }
}
